mempool/clist_sidecar.go

```diff
@@ -0,0 +1,655 @@
+
+		if e, ok := sc.txsMap.Load(tx.Key()); ok {
+			scTx := e.(*clist.CElement).Value.(*SidecarTx)
+			scTx.Senders.LoadOrStore(txInfo.SenderID, true)
+		}
+
+		return ErrTxInCache
+	}
+
+	scTx := &SidecarTx{
+		DesiredHeight: txInfo.DesiredHeight,
+		Tx:            tx,
+		BundleID:      txInfo.BundleID,
+		BundleOrder:   txInfo.BundleOrder,
+		BundleSize:    txInfo.BundleSize,
+		GasWanted:     txInfo.GasWanted,
+	}
+
+	// add to metrics that we've received a new tx
+	sc.metrics.NumMevTxsTotal.Add(1)
+
+	// -------- BASIC CHECKS ON TX INFO ---------
+
+	// Can't add transactions asking to be included in a height for auction we're not on
+	if txInfo.DesiredHeight < sc.heightForFiringAuction {
+		sc.logger.Info(
+			"failed adding sidecarTx",
+			"reason", "trying to add a tx for wrong height",
+			"desiredHeight", txInfo.DesiredHeight,
+			"currAuctionHeight", sc.heightForFiringAuction,
+			"bundleId", txInfo.BundleID,
+			"bundleOrder", txInfo.BundleOrder,
+			"tx", tx.Hash(),
+		)
+		return ErrWrongHeight{
+			DesiredHeight:        int(txInfo.DesiredHeight),
+			CurrentAuctionHeight: int(sc.heightForFiringAuction),
+		}
+	}
+
+	// revert if tx asking to be included has an order greater/equal to size
+	if txInfo.BundleOrder >= txInfo.BundleSize {
+		sc.logger.Info(
+			"failed adding sidecarTx",
+			"reason", "trying to insert a tx for bundle at an order greater than the size of the bundle...",
+			"bundleId", txInfo.BundleID,
+			"bundleOrder", txInfo.BundleOrder,
+			"bundleSize", txInfo.BundleSize,
+			"tx", tx.Hash(),
+		)
+		return ErrTxMalformedForBundle{
+			BundleID:     txInfo.BundleID,
+			BundleSize:   txInfo.BundleSize,
+			BundleHeight: txInfo.DesiredHeight,
+			BundleOrder:  txInfo.BundleOrder,
+		}
+	}
+
+	// -------- BUNDLE EXISTENCE CHECKS ---------
+
+	var bundle *Bundle
+	// load existing bundle, or MAKE NEW if not
+	existingBundle, _ := sc.bundles.LoadOrStore(Key{txInfo.DesiredHeight, txInfo.BundleID}, &Bundle{
+		DesiredHeight: txInfo.DesiredHeight,
+		BundleID:      txInfo.BundleID,
+		CurrSize:      int64(0),
+		EnforcedSize:  txInfo.BundleSize,
+		// TODO: add from gossip info?
+		GasWanted:     int64(0),
+		OrderedTxsMap: &sync.Map{},
+	})
+	bundle = existingBundle.(*Bundle)
+
+	// -------- BUNDLE SIZE CHECKS ---------
+
+	// check if bundle is asking for a different size than one already stored
+	if txInfo.BundleSize != bundle.EnforcedSize {
+		sc.logger.Info(
+			"failed adding sidecarTx",
+			"reason", "trying to insert a tx for bundle at an order greater than the size of the bundle...",
+			"bundle id", txInfo.BundleID,
+			"bundle size", txInfo.BundleSize,
+			"gasWanted", txInfo.GasWanted,
+			"enforced size", bundle.EnforcedSize,
+			"tx", tx.Hash(),
+		)
+		return ErrTxMalformedForBundle{
+			BundleID:     txInfo.BundleID,
+			BundleSize:   txInfo.BundleSize,
+			BundleHeight: txInfo.DesiredHeight,
+			BundleOrder:  txInfo.BundleOrder,
+		}
+	}
+
+	shouldReturn, err := func() (bool, error) {
```

Comment:
I think using anonymous functions makes the code more difficult to follow. I understand that this is done to use `defer` without locking the mutex up to the end of `AddTx`. Given that it is trivial to simply unlock before returning, I'd recommend in this case getting rid of the anonymous function.